scripts/release-cli.js

```diff
@@ -0,0 +1,190 @@
+
+const patchFiles = glob.sync(`patches/v4/*.patch`, {
+  cwd: path.join(__dirname, `..`),
+})
+let commitCreated = false
+
+;(async () => {
+  try {
+    // check access
+    await promiseSpawn(
+      process.execPath,
+      [`./scripts/check-publish-access/index.js`],
+      {
+        cwd: path.resolve(__dirname, `../`),
+      }
+    )
+
+    const bumpType = `major`
+    const tagName = `alpha-v4`
+    const preId = `alpha-v4`
+
+    try {
+      await Promise.all([
+        promiseSpawn(`git`, [`diff`, `--quiet`]),
+        promiseSpawn(`git`, [`diff`, `--quiet`, `--cached`]),
+      ])
+    } catch (err) {
+      throw new Error(
+        `Make sure to commit all your changes, before running a release`
+      )
+    }
+
+    // Remove all dist files so we can recompile cleanly
+    await promiseSpawn(
+      process.execPath,
+      [`./scripts/clear-package-dir.js`, bumpType],
+      {
+        cwd: path.resolve(__dirname, `../`),
+        stdio: [`inherit`, `inherit`, `inherit`],
+      }
+    )
+
+    console.log(` `)
+    console.log(`=== APPLYING GIT PATCHES ===`)
+
+    patchFiles.forEach(file => {
+      console.log(`Applying patch ${file}`)
+      try {
+        execSync(`git apply ${file}`, {
+          cwd: path.join(__dirname, `..`),
+          stdio: `pipe`,
+        })
+      } catch (err) {
+        console.log(err.stderr.toString())
+        // eslint-disable-next-line
+        throw ``
+      }
+    })
+
+    console.log(`=== COMMITING PATCH FILES ===`)
+    try {
+      await promiseSpawn(
+        `git`,
+        [`commit`, `-am`, `Comitting patch files changes`],
+        {
+          cwd: path.resolve(__dirname, `../`),
+          stdio: [`inherit`, `inherit`, `inherit`],
+        }
+      )
+      commitCreated = true
+    } catch (err) {
+      // no catch
+    }
+
+    const COMPILER_OPTIONS = `GATSBY_MAJOR=4`
+
+    console.log(` `)
+    console.log(`=== BUILDING V4 ALPHA ===`)
+    await promiseSpawn(`yarn`, [`bootstrap`], {
+      shell: true,
+      env: {
+        COMPILER_OPTIONS,
+      },
+      stdio: [`inherit`, `inherit`, `inherit`],
+    })
+
+    console.log(` `)
+    console.log(`=== PUBLISHING V4 ALPHA ===`)
+    await promiseSpawn(
+      process.execPath,
+      [
+        `./node_modules/lerna/cli.js`,
+        `publish`,
+        `--canary`,
+        `premajor`,
+        `--ignore-scripts`,
+        `--exact`,
+        `--preid`,
+        preId,
+        `--pre-dist-tag`,
+        tagName,
+        `--force-publish`, // publish all
+        `--registry`,
+        argv.registry,
+      ],
+      {
+        cwd: path.resolve(__dirname, `../`),
+        stdio: [`inherit`, `inherit`, `inherit`],
+      }
+    )
+  } catch (err) {
+    if (err) {
+      console.log(err)
+    }
+  } finally {
+    console.log(` `)
+    console.log(`=== CLEANING UP ===`)
+    if (commitCreated) {
+      console.log(`REMOVING PATCH COMMIT`)
+      try {
+        await promiseSpawn(`git`, [`reset`, `--hard`, `HEAD~1`], {
```

Comment:
whenever we use `HEAD~1` type of resets we often run into problems with script resetting to weird stuff in some cases (<cough> problems with assert-changed-files script) - if it would be easy to do, maybe we could just get current commit sha (before applying patch) and reset to that sha instead of that relative `HEAD` stuff?

Reply:
👍